Library/Homebrew/service.rb

```diff
@@ -0,0 +1,183 @@
+    end
+
+    sig { params(command: T.nilable(T.any(T::Array[String], String, Pathname))).returns(T.nilable(Array)) }
+    def run(command = nil)
+      case T.unsafe(command)
+      when nil
+        @run
+      when String, Pathname
+        @run = [command]
+      when Array
+        @run = command
+      else
+        raise TypeError, "Service#run expects an Array"
+      end
+    end
+
+    sig { params(path: T.nilable(T.any(String, Pathname))).returns(T.nilable(String)) }
+    def working_dir(path = nil)
+      case T.unsafe(path)
+      when nil
+        @working_dir
+      when String, Pathname
+        @working_dir = path.to_s
+      else
+        raise TypeError, "Service#working_dir expects a String"
+      end
+    end
+
+    sig { params(path: T.nilable(T.any(String, Pathname))).returns(T.nilable(String)) }
+    def log_path(path = nil)
+      case T.unsafe(path)
+      when nil
+        @log_path
+      when String, Pathname
+        @log_path = path.to_s
+      else
+        raise TypeError, "Service#log_path expects a String"
+      end
+    end
+
+    sig { params(path: T.nilable(T.any(String, Pathname))).returns(T.nilable(String)) }
+    def error_log_path(path = nil)
+      case T.unsafe(path)
+      when nil
+        @error_log_path
+      when String, Pathname
+        @error_log_path = path.to_s
+      else
+        raise TypeError, "Service#error_log_path expects a String"
+      end
+    end
+
+    sig { params(value: T.nilable(T::Boolean)).returns(T.nilable(T::Boolean)) }
+    def keep_alive(value = nil)
+      case T.unsafe(value)
+      when nil
+        @keep_alive
+      when true, false
+        @keep_alive = value
+      else
+        raise TypeError, "Service#keep_alive expects a Boolean"
+      end
+    end
+
+    sig { params(type: T.nilable(T.any(String, Symbol))).returns(T.nilable(String)) }
+    def run_type(type = nil)
+      case T.unsafe(type)
+      when nil
+        @run_type
+      when "immediate", :immediate
+        @run_type = type.to_s
+      when RUN_TYPE_INTERVAL, RUN_TYPE_CRON
+        raise TypeError, "Service#run_type does not support timers"
+      when String
+        raise TypeError, "Service#run_type allows: '#{RUN_TYPE_IMMEDIATE}'/'#{RUN_TYPE_INTERVAL}'/'#{RUN_TYPE_CRON}'"
+      else
+        raise TypeError, "Service#run_type expects a string"
+      end
+    end
+
+    sig { params(variables: T.nilable(T::Hash[String, String])).returns(T.nilable(T::Hash[String, String])) }
+    def environment_variables(variables = {})
+      case T.unsafe(variables)
+      when nil
+        @environment_variables
+      when Hash
+        @environment_variables = variables
+      else
+        raise TypeError, "Service#environment_variables expects a hash"
+      end
+    end
+
+    # The directory where the formula's variable files should be installed.
+    # This directory is not inside the `HOMEBREW_CELLAR` so it persists
+    # across upgrades.
+    def var
+      HOMEBREW_PREFIX/"var"
```

Comment:
`@formula.var` and same deal for the rest below. Ideally use `delegate`

Reply:
Results in:
```console
Error: Failed to load formula: appium
appium: undefined method `opt_prefix' for #<Class:0x00007fa2c7396608>
```

Reply:
Any suggestions how I can make formula child class methods accessible from the service class would be very welcome

Reply:
My suggestion would be to instead store the block and evaluate it only when the formula instance is created.

Reply:
`@formula.class.var` might work here?

Reply:
It does not, but we're back to the original issue of "Sean doesn't know enough ruby to call `@formula` methods correctly" here. If we ever want this merged (and I do), someone else needs to help me figure out the steps to setup a proper formula reference in the Service class.

Reply:
Yeah, it `@formula` is a class here rather than a formula instance. `var` etc are instance methods.

Reply:
You're not going to get a formula instance in `service do` because it doesn't exist (yet). `delegate service: :"self.class"` is part of the problem; it means that even when you do have a formula instance (rather than a class) you're throwing it away and using the class.



> My suggestion would be to instead store the block and evaluate it only when the formula instance is created, or something along those lines.

Something like this makes sense to me; only evaluate this block when you `formula.service.to_plist` is called and make use of the `formula` instance in that case to do an `instance_eval` on it to get all these `etc` methods for "free" without needing to declare them here.

Hopefully others can help with that. I can't promise I'll be able to do it soon.